### PR TITLE
waypoints.jquery: jQuery Adapter of waypoints 4.x (jquery.waypoints)

### PR DESCRIPTION
**Pull request by evil-shrike**
There's a typing in the repo for [waypoints](https://github.com/imakewebthings/waypoints) lib, but it doesn't contain jQuery Adapter. This PR adds another typing package with this adapter. As this adapter depends on jQuery it's implemented as a separated package (folder in the repo) to not impose users an additional dependency.  
Also a small refactoring was required in the main `waypoints` package: refactored `WaypointOptions` interface to allow waypoint.jquery definition - common options were extracted into base interface `WaypointOptionsBase` (as some options are optionak for usage vua jQuery).

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://imakewebthings.com/waypoints/guides/jquery-zepto/
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by andy-ms** [comment]
How is the jquery functionality for waypoints typically installed?

File: types/waypoints.jquery/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+/// <reference types="waypoints" />
+
+interface JQuery {
+	waypoint(handler: (this: Waypoint, direction?: string) => void, options?: WaypointOptionsJQuery): JQuery;
+	waypoint(options?: WaypointOptionsJQuery): JQuery;
+}
+
+interface WaypointOptionsJQuery extends WaypointOptionsBase {
+    handler?: (this: Waypoint, direction?: string) => void;
+    context?: any; // Actually it's - `HTMLElement|string` but tsc fails with error TS2430: Interface 'WaypointOptionsJQuery' incorrectly extends interface 'WaypointOptionsBase'
```
Comment (andy-ms) [inline types/waypoints.jquery/index.d.ts]: Maybe `context` shouldn't be a part of `WaypointOptionsBase` then?

Comment (evil-shrike) [reply types/waypoints.jquery/index.d.ts]: agreed, fixed.

**Comment by evil-shrike** [comment]
> How is the jquery functionality for waypoints typically installed?

it's usually used as a special build of the lib - `jquery.waypoints.js` from https://github.com/imakewebthings/waypoints/tree/master/lib (there's a build w/o jquery and there's a build with jquery).

**Comment by andy-ms** [comment]
If that file is in the `waypoints` npm package, I think this should be moved to `types/waypoints/jquery.d.ts` to be installed along with the library. People not using the jquery helpers can choose to not reference that file.

**Comment by evil-shrike** [comment]
> If that file is in the waypoints npm package, I think this should be moved to types/waypoints/jquery.d.ts to be installed along with the library. People not using the jquery helpers can choose to not reference that file.

If I move the adapter into a file inside the main package (waypoints) (e.g. "jquery.d.ts") what will happen when people install run `npm i @types/waypoints`  - will it install "jquery" package as well?

As I understand tooling generates a package.json with all dependencies found in "///<reference types" in all *.d.ts, doesn't it?

That new dependency on jquery I wanted to avoid..

**Comment by andy-ms** [comment]
You can add it to `"files"` in `tsconfig.json` to get it published. But TypeScript by default only picks up `index.d.ts` and not any other files, so you'd have to write `/// <reference types="waypoints/jquery" />` explicitly to get it.
It might still have a dependency on `@types/jquery` if your source code includes `// <reference types="jquery" />` (though it looks like that is only in the tests right now) -- I don't think it's such a bad thing to have an unused `@types` dependency though.

**Comment by evil-shrike** [comment]
Sorry by dependency I meant a npm dependency. If a package has `/// <reference types="jquery" />` and it's installed via npm, for example `npm i @types/jquery-mockjax`, npm will install a `@types/jquery` as well (as it's a dependency in generated `package.json`).

**Comment by andy-ms** [comment]
That's what I meant too. :) If your `jquery.d.ts` file includes `/// <reference types="jquery" />` then the produced packages will have a dependency on `@types/jquery`. Right now it's only in a test file though, so we wouldn't add that dependency.

**Comment by evil-shrike** [comment]
Ok, if that dep won't go into package.json then it makes sense to include adapter into the main package indeed. I'll update PR, thanks.

**Comment by evil-shrike** [comment]
I guess @typescript-bot missed pinging of the package owner as the PR was initially a "New Definition". Also it didn't remove "The Travis CI build failed" label.

**Comment by andy-ms** [comment]
Thanks!

**Pull request merged**